src/demProfiler/index.ts

```diff
@@ -0,0 +1,237 @@
+  const tileCache = new Map<number, TileImage>();
+
+  // request all the tiles we need
+  const requests: Array<Promise<TileImage | undefined>> = [];
+  for (const tile of tiles) {
+    requests.push(
+      tileRequest(tile.x, tile.y, tile.z)
+        .then((res: TileImage): TileImage => {
+          tileCache.set(tile.id, res);
+          return res;
+        })
+        .catch((err: unknown) => {
+          console.error(err);
+          return undefined;
+        })
+    );
+  }
+  // wait for all the requests to finish
+  await Promise.allSettled(requests);
+
+  return tileCache;
+}
+
+/** Get the elevation data for each coordinate */
+function getElevations(
+  coords: TileCoverCoordinates[],
+  tileCache: Map<number, TileImage>,
+  tileSize: number,
+  elevationParser?: ElevationParser
+): ElevPoint[] {
+  const points: ElevPoint[] = [];
+
+  // get the elevation data for each coordinate
+  let curDistance = 0;
+  let prevCoord: [lon: number, lat: number] | undefined;
+  for (const { tile, coordinate } of coords) {
+    const cTile = tileCache.get(tile.id);
+    if (cTile === undefined)
+      throw new Error(
+        `Missing tile ${tile.id} (${tile.x}-${tile.y}-${tile.z})`
+      );
+    const elevation = getElevation(
+      coordinate,
+      [tile.z, tile.x, tile.y],
+      tileSize,
+      cTile,
+      elevationParser
+    );
+    points.push({
+      distance: curDistance,
+      elevation,
+      coordinate,
+      tile,
+    });
+    // calculate the distance between the points
+    if (prevCoord !== undefined) {
+      curDistance += pointDistance(prevCoord, coordinate);
+    }
+    prevCoord = coordinate;
+  }
+
+  return points;
+}
+
+/** Improve the output to include metrics about the max/min/avg elevation */
+function buildOutput(points: ElevPoint[]): Output {
+  // setup variables
+  let minElevation = Infinity;
+  let maxElevation = -Infinity;
+  let totalElevation = 0;
+  // calcuate min, max, and average
+  for (const point of points) {
+    if (point.elevation < minElevation) minElevation = point.elevation;
+    if (point.elevation > maxElevation) maxElevation = point.elevation;
+    totalElevation += point.elevation;
+  }
+
+  return {
+    distance: points[points.length - 1].distance,
+    minElevation,
+    maxElevation,
+    avgElevation: totalElevation / points.length,
+    startElevation: points[0].elevation,
+    endElevation: points[points.length - 1].elevation,
+    points,
+  };
+}
+
+/**
+ * Smooth function to make the elevation appear more aesthetically
+ * pleaseing without changing the values greatly
+ */
+function smoothElevation(points: ElevPoint[]): ElevPoint[] {
+  const newPoints: ElevPoint[] = [];
+  let prevPoint: ElevPoint | undefined;
+  for (const point of points) {
+    if (prevPoint === undefined) {
+      newPoints.push(point);
+    } else {
+      const newPoint: ElevPoint = {
+        distance: point.distance,
+        elevation: (prevPoint.elevation + point.elevation) / 2,
```

Comment:
Since the new (smooth) point is averaged only by the previous one, it will actually behave like it is slighly shifted towards it. A preferable method would be that the current point is affected by both the previous and the next, possibly with a coefficient to make sure the actual value still has more weight.

something like that maybe: 
```
elevation = (prevPoint.elevation + nextPoint.elevation + point.elevation) / 3;
```

Or with more weight on the current:
```
elevation = (prevPoint.elevation + nextPoint.elevation + point.elevation * 2) / 4;
```